Build workflow utilities with CMake.

aerorahul opened the pull request:
This PR enables the utilities under `sorc/` to be built using `cmake`.  modulefiles for `hera`, `orion` and `wcoss_dell_p3` are added that load hpc-stack.

These workflow utilities were compiled on `hera` and `orion`, but not on` wcoss_dell_p3`
These workflow utilities were compiled on `macOS`, EMC Linux box (RHEL7), Docker with GNU and IntelOneAPI containers.
All utilities built as expected. They were not tested against the hand-written makefiles on the supported HPC's for reproducibility.

Minor changes in 3 Fortran source codes `syndat_qctropcy.fd/qctropcy.f`, `gfs_bufr.fd/meteorg.f`, and `fv3nc2nemsio.fd/fv3_main.f90` were required to get these utilities to be compiled with GNU Fortran.

aerorahul committed:
add cmake build capability for workflow utilities

KateFriedman-NOAA commented:
I cloned the feature/cmake branch on Hera and ran build_workflow_utils.sh without doing anything else. I got this error:

```
-bash-4.2$ pwd
/scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake/sorc
-bash-4.2$ git branch
* feature/cmake
-bash-4.2$ sh build_workflow_utils.sh
++ uname -s
+ [[ Linux == Darwin ]]
++ which readlink
+ cmd=/usr/bin/readlink
++++ /usr/bin/readlink -f -n build_workflow_utils.sh
+++ dirname /scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake/sorc/build_workflow_utils.sh
++ cd /scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake/sorc
++ pwd -P
+ readonly UTILS_DIR=/scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake/sorc
+ UTILS_DIR=/scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake/sorc
+ target=NULL
+ modulefile=/scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake/sorc/../modulefiles/workflow_utils.NULL
+ [[ -f /scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake/sorc/../modulefiles/workflow_utils.NULL ]]
+ [[ -d nceplibs-ncio ]]
+ git clone -b develop https://github.com/noaa-emc/nceplibs-ncio
Cloning into 'nceplibs-ncio'...
remote: Enumerating objects: 28, done.
remote: Counting objects: 100% (28/28), done.
remote: Compressing objects: 100% (22/22), done.
remote: Total 198 (delta 11), reused 16 (delta 6), pack-reused 170
Receiving objects: 100% (198/198), 107.73 KiB | 4.90 MiB/s, done.
Resolving deltas: 100% (102/102), done.
+ cd nceplibs-ncio
+ mkdir -p build
+ cd build
+ cmake -DCMAKE_INSTALL_PREFIX=../install
CMake Error: The source directory "/scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake/sorc/nceplibs-ncio/build" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
```

I noticed it didn't know the target (other build scripts source machine-setup.sh) so I first added "target=hera" to test and then tried adding a line to source machine-setup.sh. It got "hera" as the target either way but it still error:

```
-bash-4.2$ sh build_workflow_utils.sh
++ uname -s
+ [[ Linux == Darwin ]]
++ which readlink
+ cmd=/usr/bin/readlink
++++ /usr/bin/readlink -f -n build_workflow_utils.sh
+++ dirname /scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake/sorc/build_workflow_utils.sh
++ cd /scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake/sorc
++ pwd -P
+ readonly UTILS_DIR=/scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake/sorc
+ UTILS_DIR=/scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake/sorc
+ source ./machine-setup.sh
+ target=hera
+ modulefile=/scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake/sorc/../modulefiles/workflow_utils.hera
+ [[ -f /scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake/sorc/../modulefiles/workflow_utils.hera ]]
+ set +x

The following have been reloaded with a version change:
  1) impi/2018.4.274 => impi/2018.0.4


Currently Loaded Modules:
  1) cmake/3.16.1       4) hpc-intel/18.0.5.274   7) jasper/2.0.22  10) w3emc/2.7.3  13) bacio/2.4.1   16) nemsiogfs/2.5.3  19) bufr/11.4.0
  2) hpc/1.1.0          5) impi/2018.0.4          8) zlib/1.2.11    11) sp/2.3.3     14) w3nco/2.4.1   17) sigio/2.3.2      20) hdf5/1.10.6
  3) intel/18.0.5.274   6) hpc-impi/2018.0.4      9) png/1.6.35     12) ip/3.3.3     15) nemsio/2.5.2  18) g2/3.4.1         21) netcdf/4.7.4



+ [[ -d nceplibs-ncio ]]
+ rm -rf nceplibs-ncio
+ git clone -b develop https://github.com/noaa-emc/nceplibs-ncio
Cloning into 'nceplibs-ncio'...
remote: Enumerating objects: 28, done.
remote: Counting objects: 100% (28/28), done.
remote: Compressing objects: 100% (22/22), done.
remote: Total 198 (delta 11), reused 16 (delta 6), pack-reused 170
Receiving objects: 100% (198/198), 107.73 KiB | 5.67 MiB/s, done.
Resolving deltas: 100% (102/102), done.
+ cd nceplibs-ncio
+ mkdir -p build
+ cd build
+ cmake -DCMAKE_INSTALL_PREFIX=../install
CMake Warning:
  No source or binary directory provided.  Both will be assumed to be the
  same as the current working directory, but note that this warning will
  become a fatal error in future CMake releases.


CMake Error: The source directory "/scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake/sorc/nceplibs-ncio/build" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
```

I'm not experienced with cmake...am I missing something? Thanks!

aerorahul committed:
update .gitignore to exclude build and install directories as well as compiled files. bugfix in build_workflow_utils.sh hack. nceplibs-ncio now creates the module ncio and not fv3gfs_ncio.

aerorahul commented:
@KateFriedman-NOAA 
My apologies.  The hack was missing a path to the source directory for nceplibs-ncio.
It has been fixed. I tested on Orion:
`target=orion ./build_workflow_utils.sh`

KateFriedman-NOAA commented:
@aerorahul Thanks for the fix! I updated my clone of feature/cmake on Hera and all workflow util codes built, woohoo! 

> These workflow utilities were compiled on hera and orion, but not on wcoss_dell_p3

Question...why not on wcoss_dell_p3? Are we waiting for a library update there?

aerorahul commented:
I did not build on wcoss_dell_p3. I was lazy. But, it should build.
When hpc-stack gets `nceplibs-ncio` installed on all machines within hpc-stack, we can remove the hack that is in `build_workflow_utils.sh` and uncomment the line in `modulefiles/workflow_utils.<target>` that loads the `ncio` module.

KateFriedman-NOAA commented:
@aerorahul Gotcha, I'll try the build on WCOSS-Dell and report any issues.

I ran a cycle test overnight on Hera with feature/cmake. The gdasechgres job segfaulted:

/scratch1/NCEPDEV/stmp4/Kate.Friedman/comrot/cmake/logs/2020090118/gdasechgres.log.1

I reran with module_base.hera using hpc-stack to test that but it segfaulted similarly:

/scratch1/NCEPDEV/stmp4/Kate.Friedman/comrot/cmake/logs/2020090118/gdasechgres.log

I see mention of module_ncio, do we need to load an additional module at runtime? Thanks!

Also, there is some additional updating/cleanup/decommissioning work to be done on feature/cmake (e.g. remove unneeded build scripts). I had to do some updating to get the branch ready to run so I'll finish that and commit those updates.

aerorahul commented:
@KateFriedman-NOAA 
`module_ncio` is a fortran module and not a modulefile.
I will investigate the failure.

The cleanup/decommissioning should be done in a separate branch and not in `feature/cmake` and in a follow up PR.
You can push changes that you need to make the branch ready to run.

aerorahul commented:
@KateFriedman-NOAA 
I suspect there is an issue in the `ipolates` routine coming from `ip/3.3.3`.
Can you confirm that your branch `feature/hpc-stack` works on hera for `enkf_chgres_recenter_nc.fd`?
I will tag @GeorgeGayno-NOAA and @kgerheiser as they have been in the weeds of `ip` lately.

aerorahul committed:
copy/paste error from enkf_chgres_recenter.fd to enkf_chgres_recenter_nc.fd.  When going from nemsio to netcdf, the linking of ip, sp and w3nco changed from _d to _4.

aerorahul commented:
@KateFriedman-NOAA 
I found the copy/paste error from `enkf_chgres_recenter.fd` to `enkf_chgres_recenter_nc.fd` in `CMakeLists.txt`.  When going from `nemsio` to `netcdf`, the linking of `ip`, `sp` and `w3nco` changed from `_d` to `_4`.
I tested the program and it now terminates normally.

KateFriedman-NOAA commented:
@aerorahul Thanks for the fix! I updated my clone on Hera, reran build_workflow_utils.sh and then rewound the failed gdasechgres job. The job completed successfully!

/scratch1/NCEPDEV/stmp4/Kate.Friedman/comrot/cmake/logs/2020090118/gdasechgres.log

I also ran the gdasechgres job in my feature/hpc-stack test (symlinked workflow execs to feature/cmake execs) and it also completed successfully:

/scratch1/NCEPDEV/stmp4/Kate.Friedman/comrot/stacktest1/logs/2020090118/gdasechgres.log

I am letting my cmake test on Hera continue forward another cycle. I want to run a cycle on both Orion and WCOSS-Dell, and repeat the test on Hera for a reproducibility check.

I'll have a commit shortly to add workflow_utils to the build_all and linking scripts so it builds/links out of the box again. Further cleanup and decommissioning will happen in a follow-up PR.

KateFriedman-NOAA committed:
Add workflow_utils to build_all and link scripts

- add build_workflow_utils.sh to build_all.sh
- add workflow utils exec symlinks to link_fv3gfs.sh since these execs no longer land in /exec folder
- add workflow_utils build to partial_build.sh and fv3gfs_build.cfg for use by build_all
- turn off builds in fv3gfs_build.cfg that are now covered by workflow_utils build
- move machine-setup source in build_workflow_utils.sh before target is needed

Further cleanup and decommissioning of consolidated builds will come in follow-up commit/PR.

Refs: #264

aerorahul commented on sorc/build_workflow_utils.sh:
```diff
@@ -5,11 +5,11 @@ set -eux
 readonly UTILS_DIR=$(cd "$(dirname "$($cmd -f -n "${BASH_SOURCE[0]}" )" )" && pwd -P)
 
 # Adapt for global-workflow structure.
+source ${UTILS_DIR}/machine-setup.sh > /dev/null 2>&1
```
can you move this line inside the if-block.
There is no such file when building on a generic platform.
This is specific to how we do things in our modulefiles.

aerorahul replied:
specifically after `set +x`.
The gibberish output from module commands makes no sense for the user to see.

KateFriedman-NOAA replied:
So our convention currently is not to provide $target as the build command is done (e.g. target=hera ./build_workflow_util.sh) but to provide it from the top of build_all.sh (source machine-setup.sh) and/or within the build script itself (which allows it to be run by itself outside of build_all.sh). Because of that we need $target set before the modulefile line otherwise the following happens and it never makes it into the if-block to run machine-setup and update $modulefile:

```
-bash-4.2$ sh build_workflow_utils.sh
++ uname -s
+ [[ Linux == Darwin ]]
++ which readlink
+ cmd=/usr/bin/readlink
++++ /usr/bin/readlink -f -n build_workflow_utils.sh
+++ dirname /scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake_test/sorc/build_workflow_utils.sh
++ cd /scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake_test/sorc
++ pwd -P
+ readonly UTILS_DIR=/scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake_test/sorc
+ UTILS_DIR=/scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake_test/sorc
+ target=NULL
+ modulefile=/scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake_test/sorc/../modulefiles/workflow_utils.NULL
+ [[ -f /scratch1/NCEPDEV/global/Kate.Friedman/git/feature-cmake_test/sorc/../modulefiles/workflow_utils.NULL ]]
```

aerorahul replied:
You do set target in `build_all.sh`, why repeat that in every single build script?
You can do this:
target=$target ./build_workflow_utils.sh

aerorahul replied:
doing so, you are breaking the `build_workflow_utils.sh` for generic machines that don't have/need modules.

KateFriedman-NOAA replied:
> You can do this:
> target=$target ./build_workflow_utils.sh

That's perfectly fine for build_all.sh but then requires users to also provide $target like that when running that build script alone. That goes against the current convention...but I'll revert my change to build_workflow_utils.sh, add $target to build_all.sh for that build script and call it a day. Can come back to reworking the build scripts later.

aerorahul commented on sorc/fv3gfs_build.cfg:
```diff
@@ -9,14 +9,15 @@
- Building gfs_bufrsnd (gfs_bufrsnd) .................... yes
- Building fv3nc2nemsio (fv3nc2nemsio) .................. yes
- Building regrid_nemsio (regrid_nemsio) ................ yes
+ Building gaussian_sfcanl (gaussian_sfcanl)............. no
+ Building enkf_chgres_recenter (enkf_chgres_recenter) .. no
+ Building enkf_chgres_recenter_nc (enkf_chgres_recenter_nc) .. no
+ Building tropcy_NEMS (tropcy) ......................... no
+ Building gfs_fbwndgfs (gfs_fbwndgfs) .................. no
+ Building gfs_bufrsnd (gfs_bufrsnd) .................... no
+ Building fv3nc2nemsio (fv3nc2nemsio) .................. no
+ Building regrid_nemsio (regrid_nemsio) ................ no
```
why no? we are building them as part of `workflow_utils`.

aerorahul replied:
never mind. 
I suppose this will be cleaned up later.

KateFriedman-NOAA replied:
The ones set to "no" will be removed during the cleanup/decommissioning stage.

aerorahul commented on sorc/build_all.sh:
```diff
@@ -266,6 +266,20 @@ fi
+#------------------------------------
+$Build_workflow_utils && {
+echo " .... Building workflow_utils .... "
+./build_workflow_utils.sh > $logs_dir/build_workflow_utils.log 2>&1
```
why is this not acceptable?
```
target=$target ./build_workflow_utils.sh > $logs_dir/build_workflow_utils.log 2>&1
```

KateFriedman-NOAA committed:
Adjust how target is set for build_workflow_utils

- Add $target to build_workflow_utils.sh command in build_all.sh
- Move machine-setup source back down into if-block in build_workflow_utils.sh

Refs: #264

KateFriedman-NOAA commented:
@aerorahul The analcalc jobs are failing in my test of feature/cmake on Hera. Can you take a look? Thanks!

/scratch1/NCEPDEV/stmp4/Kate.Friedman/comrot/cmake/logs/2020090200/gdasanalcalc.log
/scratch1/NCEPDEV/stmp4/Kate.Friedman/comrot/cmake/logs/2020090200/gfsanalcalc.log

aerorahul committed:
another use of _d where an _4 is needed

aerorahul commented:
@WalterKolczynski-NOAA @KateFriedman-NOAA 
Can you as part of the review check and make sure that all the compiler flags and library linking was accurately transferred from the `Makefile` to `CMakeLists.txt`.

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA @KateFriedman-NOAA
> Can you as part of the review check and make sure that all the compiler flags and library linking was accurately transferred from the `Makefile` to `CMakeLists.txt`.

Will do

WalterKolczynski-NOAA commented on sorc/fv3nc2nemsio.fd/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+)
+
+if(CMAKE_Fortran_COMPILER_ID MATCHES "^(Intel)$")
+  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -fp-model precise")
```
There seem to be two version of the build script for this currently, but neither seem to use the '-fp-model precise' flag

aerorahul replied:
I can remove it.  Thanks for spotting this.

aerorahul replied:
done.

WalterKolczynski-NOAA commented on sorc/fv3nc2nemsio.fd/fv3_main.f90:
```diff
@@ -63,13 +63,13 @@ program fv3_main
     allocate (name3din(nvar3d), name3dout(nvar3d))
-    name3din=(/'ucomp','vcomp','temp','sphum','o3mr','nhpres','w','clwmr','delp'/)
-    name3dout=(/'ugrd','vgrd','tmp','spfh','o3mr','pres','vvel','clwmr','dpres'/)
+    name3din=(/'ucomp ','vcomp ','temp  ','sphum ','o3mr  ','nhpres','w     ','clwmr ','delp  '/)
```
Any reason for the added spaces?

aerorahul replied:
GNU Fortran requires character arrays to be of same length.
Intel Fortran does not care.

WalterKolczynski-NOAA commented on modulefiles/workflow_utils.hera:
```diff
@@ -0,0 +1,29 @@
+module load hpc-intel/18.0.5.274
+module load hpc-impi/2018.0.4
+
+module load jasper/2.0.22
```
Update to the fixed jasper version 2.0.25

aerorahul replied:
Is that installed in hpc-stack/1.1.0?  If so, I can update.

aerorahul replied:
done.

WalterKolczynski-NOAA commented on modulefiles/workflow_utils.orion:
```diff
@@ -0,0 +1,29 @@
+module load hpc-intel/2018.4
+module load hpc-impi/2018.4
+
+module load jasper/2.0.22
```
Update to the fixed jasper version 2.0.25

aerorahul replied:
see above comment.

aerorahul replied:
done

WalterKolczynski-NOAA commented on modulefiles/workflow_utils.wcoss_dell_p3:
```diff
@@ -0,0 +1,29 @@
+module load hpc-ips/18.0.1.163
+module load hpc-impi/18.0.1
+
+module load jasper/2.0.22
```
Update to the fixed jasper version 2.0.25

aerorahul replied:
see above comment.

aerorahul replied:
done

WalterKolczynski-NOAA commented on sorc/enkf_chgres_recenter.fd/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+  sp::sp_d
+  w3nco::w3nco_d)
+if(OpenMP_Fortran_FOUND)
+  target_link_libraries(${exe_name} OpenMP::OpenMP_Fortran)
```
Will this take care of adding the -qopenmp flag?

aerorahul replied:
yes.

WalterKolczynski-NOAA commented on sorc/enkf_chgres_recenter_nc.fd/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+  sp::sp_4
+  w3nco::w3nco_4
+  ncio::ncio
+  NetCDF::NetCDF_Fortran)
```
I'm assuming NetCDF picks up hdf5 and z libs as well.

aerorahul replied:
yes.

WalterKolczynski-NOAA commented on sorc/enkf_chgres_recenter_nc.fd/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+  ncio::ncio
+  NetCDF::NetCDF_Fortran)
+if(OpenMP_Fortran_FOUND)
+  target_link_libraries(${exe_name} OpenMP::OpenMP_Fortran)
```
Same question as before about -qopenmp flag.

aerorahul replied:
same answer. yes.

WalterKolczynski-NOAA commented on sorc/build_workflow_utils.sh:
```diff
@@ -0,0 +1,48 @@
+# After nceplibs-ncio is in hpc-stack, add the following line to
+# ${UTILS_DIR}/../modulefiles/workflow_utils.<platform>
+# "module load ncio/<ncio-version>"
+# and remove this hack
```
ncio looks to be available now in hpc-stack as ncio/1.0.0

aerorahul replied:
it is an unofficial release.  Ed is working on adding documentation and tests.  When it will be released, we can remove the hack.
This will ensure we work with the current develop leading up to the release.

WalterKolczynski-NOAA replied:
Got it.

WalterKolczynski-NOAA commented on sorc/gfs_bufr.fd/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+  bufr::bufr_4_DA
+  NetCDF::NetCDF_Fortran)
+if(OpenMP_Fortran_FOUND)
+  target_link_libraries(${exe_name} OpenMP::OpenMP_Fortran)
```
As before on -qopenmp

aerorahul replied:
yes.

WalterKolczynski-NOAA commented on sorc/gfs_bufr.fd/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+)
+
+if(CMAKE_Fortran_COMPILER_ID MATCHES "^(Intel)$")
+  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -convert big_endian -fp-model source")
```
Neither -convert big_endian nor -fp-model source seem to be set in the existing makefile (gfs_bufr.fd/makefile_module)

aerorahul replied:
It is defined in the modulefile.
I am not sure why and who thought setting compiler flags in a modulefile was a great idea.

WalterKolczynski-NOAA commented on sorc/regrid_nemsio.fd/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+)
+
+if(CMAKE_Fortran_COMPILER_ID MATCHES "^(Intel)$")
+  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -fp-model precise")
```
-fp-model precise flag not used in existing makefile (regrid_nemsio.fd/Makefile)

aerorahul replied:
you are right.  The flags are set in the build script.  This flag is not present.  I will remove.

KateFriedman-NOAA replied:
Was this resolved?

aerorahul replied:
done.

WalterKolczynski-NOAA commented on sorc/regrid_nemsio.fd/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+  sp::sp_d
+  w3nco::w3nco_d
+  NetCDF::NetCDF_Fortran
+  MPI::MPI_Fortran)
```
MPI library doest not seem to be linking in current makefile, yet it clearly belongs.

aerorahul replied:
We typically define `FC=mpif90` instead of `FC=ifort`.
For a program `mpif90` will take care of linking the appropriate mpi library when building.
When `ifort` is used, it will require the necessary MPI libraries to be linked at compile time.
There is no "rule" to this.
In the current sorc, there are serial programs and MPI programs.  It is "unfair" to compile the serial programs with `mpif90`.  This explicit linking takes care of it and allows building serial and parallel programs as they should.

WalterKolczynski-NOAA commented on sorc/supvit.fd/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+)
+
+if(CMAKE_Fortran_COMPILER_ID MATCHES "^(Intel)$")
+  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -i4 -r8 -assume byterecl -assume noold_ldout_format")
```
The two -assume flags do not appear to be in the existing makefile (supvit.fd/makefile)

aerorahul replied:
you are correct.  I can remove.

KateFriedman-NOAA replied:
was this resolved?

aerorahul replied:
done.

WalterKolczynski-NOAA commented on sorc/syndat_maksynrc.fd/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+add_executable(${exe_name} ${fortran_src})
+target_link_libraries(
+  ${exe_name}
+  w3nco::w3nco_4)
```
Current build script (sorc/build_tropcy_NEMS.sh) also includes BACIO_LIB4. Possibly pulled in as a dependency by w3nco?

aerorahul replied:
True.  But there is no call to any bacio library function.  I think it might have been an error.
The build would have failed if the program did require it.

WalterKolczynski-NOAA commented on sorc/tave.fd/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+  ${exe_name}
+  bacio::bacio_4
+  w3nco::w3nco_d
+  g2::g2_d)
```
Current build script (sorc/build_tropcy_NEMS.sh) also links W3EMC_LIBd, PNG_LIB, JASPER_LIB, and Z_LIB.

aerorahul replied:
PNG, Jasper and ZLib are dependencies of g2.
When looking for g2, cmake ensures its (g2's) dependencies are found and included.
By large in cmake, one should only need to link to dependencies that the program/library uses.  It is the job of the dependencies to provide its own dependencies.

WalterKolczynski-NOAA replied:
That's what I expected. What about w3emc_libd? (I still assume it would've failed to build if needed.)

aerorahul replied:
I suspect it is not needed since I didn't see any calls to functions from that library.
If there was a test for this program (or any program for that matter), it would be easy to find out.

WalterKolczynski-NOAA commented on sorc/tocsbufr.fd/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+)
+
+if(CMAKE_Fortran_COMPILER_ID MATCHES "^(Intel)$")
+  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -convert big_endian -fp-model source")
```
Neither -convert big_endian nor -fp-model source seem to be set in the existing makefile (tocsbufr.fd/makefile_module)

aerorahul replied:
`modulefiles/gfs_bufr.hera` defines it and it is loaded when building this program.

WalterKolczynski-NOAA commented on sorc/vint.fd/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+  ${exe_name}
+  bacio::bacio_4
+  w3nco::w3nco_d
+  g2::g2_d)
```
Current build script (sorc/build_tropcy_NEMS.sh) also links W3EMC_LIBd, PNG_LIB, JASPER_LIB, and Z_LIB.

aerorahul replied:
see above comment on how cmake handles dependencies.

aerorahul replied:
see above comment on how cmake handles dependencies.

WalterKolczynski-NOAA commented on sorc/CMakeLists.txt:
```diff
@@ -0,0 +1,72 @@
+# Set compiler flags.
+if(CMAKE_Fortran_COMPILER_ID MATCHES "^(Intel)$")
+  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -g -traceback")
+  set(CMAKE_Fortran_FLAGS_RELEASE "-O3")
```
Programs seem to vary on whether to use O2 or O3 optimization, and also on whether traceback is turned on for release code (most do).

aerorahul replied:
traceback is benign and does not affect optimization.  It is very useful when something fails and hence should always be ON

WalterKolczynski-NOAA reviewed (changes requested):
Did what I could to cross-check against existing makefiles/build scripts with my rudimentary knowledge of cmake and the library dependencies. I suspect most of my comments will be either covered by cmake picking up library dependencies or just inconsequential changes. It's also possible I missed a flag being set somewhere, though I tried to check both the build script and the makefile.

aerorahul commented:
@WalterKolczynski-NOAA 
Thank you for a careful review and cross-examination of compiler flags and dependencies.
As you can see, the ingredients are sprinkled across files and there is no "convention" or "consistency".
Please examine the responses and mark the ones that you are satisfied as "resolved".
I will work on what is left and request a re-review.

aerorahul committed:
update jasper to 2.x.25. remove compiler flag in fv3nc2nemsio

WalterKolczynski-NOAA commented:
Sorry for the hold-up. I've resolved all the ones that are definitely taken care of. I will leave it to your discretion about what to do with the remainder and whether they are impactful or not.

aerorahul commented:
@WalterKolczynski-NOAA Thanks again for the careful review.
There are no more changes from me on this PR.

WalterKolczynski-NOAA commented:
@KateFriedman-NOAA Go ahead and merge if you are satisfied.

aerorahul committed:
remove flags that were not present in Makefiles

KateFriedman-NOAA reviewed (approved):
Thanks for cmake-ifying these codes @aerorahul ! Tested branch on Hera, Orion, and WCOSS-Dell. Ran fine. Duplicate run on Hera reproduced output.

KateFriedman-NOAA commented:
Reran build_workflow_utils.sh on WCOSS-Dell, Hera, and Orion after commit at 1d7776d. Build ran fine.
